File: wizhorse/workers/ghidra_scripts/wh_decompile_function.java
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Set;
import java.util.TreeSet;

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceIterator;
import ghidra.program.model.symbol.ReferenceManager;

public class wh_decompile_function extends GhidraScript {
    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        String outputPath = args[0];
        String addressText = args[1];
        Address address = currentProgram.getAddressFactory().getAddress(addressText);
        if (address == null) {
            throw new IllegalArgumentException("invalid address: " + addressText);
        }

        Function function = currentProgram.getFunctionManager().getFunctionAt(address);
        if (function == null) {
            function = currentProgram.getFunctionManager().getFunctionContaining(address);
        }
        if (function == null) {
            throw new IllegalArgumentException("no function found at or containing address: " + addressText);
        }

        DecompInterface decompiler = new DecompInterface();
        decompiler.openProgram(currentProgram);
        DecompileResults result = decompiler.decompileFunction(function, 30, monitor);
        String pseudocode = "";
        if (result != null && result.decompileCompleted() && result.getDecompiledFunction() != null) {
            pseudocode = result.getDecompiledFunction().getC();
        }
        decompiler.dispose();

        String payload = "{"
            + "\"address\":" + quote(function.getEntryPoint().toString()) + ","
            + "\"name\":" + quote(function.getName()) + ","
            + "\"pseudocode\":" + quote(pseudocode) + ","
            + "\"callers\":" + stringArray(callers(function)) + ","
            + "\"callees\":" + stringArray(callees(function)) + ","
            + "\"referenced_strings\":" + stringArray(referencedStrings(function))
            + "}";
        Files.write(Paths.get(outputPath), payload.getBytes(StandardCharsets.UTF_8));
    }

    private Set<String> callers(Function function) {
        ReferenceManager referenceManager = currentProgram.getReferenceManager();
        Set<String> callers = new TreeSet<>();
        ReferenceIterator references = referenceManager.getReferencesTo(function.getEntryPoint());
        for (Reference reference : references) {
            Function caller = currentProgram.getFunctionManager()
                .getFunctionContaining(reference.getFromAddress());
            callers.add(caller != null ? caller.getEntryPoint().toString() : reference.getFromAddress().toString());
        }
        return callers;
    }

    private Set<String> callees(Function function) {
        ReferenceManager referenceManager = currentProgram.getReferenceManager();
        Set<String> callees = new TreeSet<>();
        InstructionIterator instructions = currentProgram.getListing()
            .getInstructions(function.getBody(), true);
        for (Instruction instruction : instructions) {
            for (Reference reference : referenceManager.getReferencesFrom(instruction.getAddress())) {
                Function target = currentProgram.getFunctionManager().getFunctionAt(reference.getToAddress());
                if (target == null) {
                    target = currentProgram.getFunctionManager().getFunctionContaining(reference.getToAddress());
                }
                if (target != null && !target.getEntryPoint().equals(function.getEntryPoint())) {
                    callees.add(target.getEntryPoint().toString());
                }
            }
        }
        return callees;
    }

    private Set<String> referencedStrings(Function function) {
        ReferenceManager referenceManager = currentProgram.getReferenceManager();
        Set<String> strings = new TreeSet<>();
        InstructionIterator instructions = currentProgram.getListing()
            .getInstructions(function.getBody(), true);
        for (Instruction instruction : instructions) {
            for (Reference reference : referenceManager.getReferencesFrom(instruction.getAddress())) {
                Data data = currentProgram.getListing().getDataContaining(reference.getToAddress());
                if (data == null) {
                    continue;
                }
                Object value = data.getValue();
                if (value == null) {
                    continue;
                }
                String text = value.toString();
                if (looksLikeString(text)) {
                    strings.add(text);
                }
            }
        }
        return strings;
    }

    private boolean looksLikeString(String text) {
        if (text.length() < 4) {
            return false;
        }
        int printable = 0;
        for (int index = 0; index < text.length(); index++) {
            char character = text.charAt(index);
            if (character >= 32 && character <= 126) {
                printable++;
            }
        }
        return printable >= Math.max(4, (int) (text.length() * 0.8));
    }

    private String stringArray(Set<String> values) {
        StringBuilder builder = new StringBuilder("[");
        boolean first = true;
        int count = 0;
        for (String value : values) {
            if (count >= 200) {
                break;
            }
            if (!first) {
                builder.append(",");
            }
            first = false;
            builder.append(quote(value));
            count++;
        }
        builder.append("]");
        return builder.toString();
    }

    private String quote(String value) {
        if (value == null) {
            return "null";
        }
        StringBuilder builder = new StringBuilder("\"");
        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);
            switch (character) {
                case '\\':
                    builder.append("\\\\");
                    break;
                case '"':
                    builder.append("\\\"");
                    break;
                case '\n':
                    builder.append("\\n");
                    break;
                case '\r':
                    builder.append("\\r");
                    break;
                case '\t':
                    builder.append("\\t");
                    break;
                default:
                    if (character < 0x20) {
                        builder.append(String.format("\\u%04x", (int) character));
                    }
                    else {
                        builder.append(character);
                    }
                    break;
            }
        }
        builder.append('"');
        return builder.toString();
    }
}

File: wizhorse/workers/ghidra_scripts/wh_decompile_function.py
import json

from ghidra.app.decompiler import DecompInterface


def main():
    args = getScriptArgs()
    output_path = args[0]
    address_text = args[1]

    address = currentProgram.getAddressFactory().getAddress(address_text)
    function_manager = currentProgram.getFunctionManager()
    function = function_manager.getFunctionAt(address)
    if function is None:
        function = function_manager.getFunctionContaining(address)

    if function is None:
        raise ValueError("no function found at or containing address: " + address_text)

    decompiler = DecompInterface()
    decompiler.openProgram(currentProgram)
    decompile_result = decompiler.decompileFunction(function, 30, monitor)
    pseudocode = ""
    if decompile_result and decompile_result.decompileCompleted():
        decompiled_function = decompile_result.getDecompiledFunction()
        if decompiled_function is not None:
            pseudocode = decompiled_function.getC()

    payload = {
        "address": str(function.getEntryPoint()),
        "name": function.getName(),
        "pseudocode": pseudocode,
        "callers": _callers(function),
        "callees": _callees(function),
        "referenced_strings": _referenced_strings(function),
    }
    with open(output_path, "w") as output_file:
        json.dump(payload, output_file)


def _callers(function):
    reference_manager = currentProgram.getReferenceManager()
    function_manager = currentProgram.getFunctionManager()
    callers = set()
    for reference in reference_manager.getReferencesTo(function.getEntryPoint()):
        caller = function_manager.getFunctionContaining(reference.getFromAddress())
        if caller is not None:
            callers.add(str(caller.getEntryPoint()))
        else:
            callers.add(str(reference.getFromAddress()))
    return sorted(callers)


def _callees(function):
    listing = currentProgram.getListing()
    reference_manager = currentProgram.getReferenceManager()
    function_manager = currentProgram.getFunctionManager()
    callees = set()
    instructions = listing.getInstructions(function.getBody(), True)
    for instruction in instructions:
        for reference in reference_manager.getReferencesFrom(instruction.getAddress()):
            target = function_manager.getFunctionAt(reference.getToAddress())
            if target is None:
                target = function_manager.getFunctionContaining(reference.getToAddress())
            if target is not None and target.getEntryPoint() != function.getEntryPoint():
                callees.add(str(target.getEntryPoint()))
    return sorted(callees)


def _referenced_strings(function):
    listing = currentProgram.getListing()
    reference_manager = currentProgram.getReferenceManager()
    strings = set()
    instructions = listing.getInstructions(function.getBody(), True)
    for instruction in instructions:
        for reference in reference_manager.getReferencesFrom(instruction.getAddress()):
            data = listing.getDataContaining(reference.getToAddress())
            if data is None:
                continue
            try:
                value = data.getValue()
            except Exception:
                continue
            if value is None:
                continue
            text = str(value)
            if text and _looks_like_string(text):
                strings.add(text)
    return sorted(strings)[:200]


def _looks_like_string(text):
    if len(text) < 4:
        return False
    printable = sum(1 for char in text if 32 <= ord(char) <= 126)
    return printable >= max(4, int(len(text) * 0.8))


main()
